e2e/testcases/otel_collector_test.go

```diff
@@ -0,0 +1,150 @@
+	nomostest.DeletePodByLabel(nt, "app", "opentelemetry", false)
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := validateOtelCollectorStatusCurrent(nt, false); err != nil {
+		nt.T.Fatal()
+	}
+
+	nt.T.Log("Wait 1 minute for metric exporter to settle")
+	time.Sleep(1 * time.Minute)
+	nt.T.Log("Checking the otel-collector log contains no failure...")
+	err := CheckDeploymentLogHasNoFailure(nt, ocmetrics.OtelCollectorName, ocmetrics.MonitoringNamespace)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("Apply custom otel-collector ConfigMap that could cause duplicate time series error")
+	nt.MustKubectl("apply", "-f", "../testdata/otel-collector/otel-cm-duplicate-timeseries.yaml")
+	nt.T.Log("Restart otel-collector pod to refresh the ConfigMap and log")
+	nomostest.DeletePodByLabel(nt, "app", "opentelemetry", false)
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := validateOtelCollectorStatusCurrent(nt, true); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("wait 1 minute for metric exporter to settle")
+	time.Sleep(1 * time.Minute)
+	nt.T.Log("Checking the otel-collector log contains failure...")
+	if err := CheckDeploymentLogHasFailure(nt, ocmetrics.OtelCollectorName, ocmetrics.MonitoringNamespace); err != nil {
+		nt.T.Fatal(err)
+	}
+}
+
+func CheckDeploymentLogHasFailure(nt *nomostest.NT, deployment, namespace string) error {
+	nt.T.Helper()
+
+	args := []string{"logs", fmt.Sprintf("deployment/%s", deployment), "-n", namespace}
+	cmd := fmt.Sprintf("kubectl %s", strings.Join(args, " "))
+	out, err := nt.Shell.Kubectl(args...)
+	if err != nil {
+		nt.T.Logf("failed to run %q: %v\n%s", cmd, err, out)
+		return err
+	}
+
+	entry := strings.Split(string(out), "\n")
+	for _, m := range entry {
+		if strings.Contains(m, "failed to export time series to GCM") {
+			return nil
+		}
+	}
+	return fmt.Errorf("error expected in the log of deployment %s, namespace %s but found none", deployment, namespace)
+}
+
+func CheckDeploymentLogHasNoFailure(nt *nomostest.NT, deployment, namespace string) error {
+	nt.T.Helper()
+
+	args := []string{"logs", fmt.Sprintf("deployment/%s", deployment), "-n", namespace}
+	cmd := fmt.Sprintf("kubectl %s", strings.Join(args, " "))
+	out, err := nt.Shell.Kubectl(args...)
+	if err != nil {
+		nt.T.Logf("failed to run %q: %v\n%s", cmd, err, out)
+		return err
+	}
+
+	entry := strings.Split(string(out), "\n")
+	for _, m := range entry {
+		if strings.Contains(m, "failed to export time series to GCM") {
+			return fmt.Errorf("failure found in the log of deployment %s, namespace %s: %s", deployment, namespace, m)
+		}
+	}
+	return nil
+}
+
+// func validateDeploymentstatuscurrent
+func validateOtelCollectorStatusCurrent(nt *nomostest.NT, customConfigMap bool) error {
+	tg := taskgroup.New()
+	tg.Go(func() error {
+		return nt.Watcher.WatchForCurrentStatus(kinds.Deployment(), ocmetrics.OtelCollectorName, ocmetrics.MonitoringNamespace)
```

Comment:
Should we maybe wait for the Deployment to be current after all the ConfigMaps are current? That way we know it has the latest ConfigMap?

Do ConfigMaps even have status? What exactly are we waiting for here?

Reply:
For the ConfigMap I just want to check it's existence.

It's a good point to wait for the ConfigMap before deployment. For the test ConfigMap (the fault one), it's applied by mustKubectl, could it provide some guarantee that the resource will be applied otherwise error out?

Reply:
I think after applying a ConfigMap, as long as the apply didn't error, you don't need to check that it's current.

However, I would recommend using the KubeClient instead of MustKubectl. The KubeClient will give is much better errors that we can catch if necessary.

Reply:
The problem with mutable ConfigMaps is that there's no way to know after updating them that the Pods have actually picked up the change. You might want to patch the otel-collector Deployment to add a spec.template.spec.metadata.label with a hash of the ConfigMap. That way you can check if the label has been updated on the pod to ensure it's been replaced since the ConfigMap was updated.

Reply:
Alternatively, you could just change the Deployment pod spec to point to a new ConfigMap (new name). That's the immutable ConfigMap strategy.

Reply:
The test ConfigMap is with a different name (using otel-collector-custom). Problem is otel-collector deployment is using the projected volume for the CMs and the bottom one will take precedence, and the ConfigMap would just overrides the config data. For now there is no way but just assuming if both `otel-collcetor-googlecloud` and `otel-collector-custom` exist under monitoring namespace, the later one will be effective according to k8s design. https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/manifests/templates/otel-collector.yaml#L163